velox/functions/prestosql/Zip.cpp

```diff
@@ -0,0 +1,221 @@
+
+    for (int i = 0; i < numInputArrays; i++) {
+      baseVectors[i] = decodedArgs.at(i)->base()->as<ArrayVector>();
+      rawSizes[i] = baseVectors[i]->rawSizes();
+      rawOffsets[i] = baseVectors[i]->rawOffsets();
+      indices[i] = decodedArgs.at(i)->indices();
+    }
+
+    // Size of elements in result vector.
+    vector_size_t resultElementsSize = 0;
+
+    // Determine what the size of the resultant elements will be so we can
+    // reserve enough space.
+    auto getMaxArraySize = [&](vector_size_t row) -> vector_size_t {
+      vector_size_t maxSize = 0;
+      for (int i = 0; i < numInputArrays; i++) {
+        maxSize = std::max(maxSize, rawSizes[i][indices[i][row]]);
+      }
+      return maxSize;
+    };
+
+    BufferPtr maxArraySizesBuffer =
+        AlignedBuffer::allocate<vector_size_t>(rows.end(), context->pool());
+    auto maxArraySizes = maxArraySizesBuffer->asMutable<vector_size_t>();
+    rows.applyToSelected([&](auto row) {
+      auto maxSize = getMaxArraySize(row);
+      resultElementsSize += maxSize;
+      maxArraySizes[row] = maxSize;
+    });
+
+    // Now create individual result vectors for each input Array vector.
+
+    auto pool = context->pool();
+    std::vector<BufferPtr> resultIndices(numInputArrays);
+    std::vector<BufferPtr> resultNulls(numInputArrays);
+    std::vector<vector_size_t*> rawResultIndices(numInputArrays);
+    std::vector<uint64_t*> rawResultNulls(numInputArrays);
+
+    for (int i = 0; i < numInputArrays; i++) {
+      resultIndices[i] = allocateIndices(resultElementsSize, pool);
+      resultNulls[i] = AlignedBuffer::allocate<bool>(
+          resultElementsSize, pool, bits::kNotNull);
+      rawResultIndices[i] = resultIndices[i]->asMutable<vector_size_t>();
+      rawResultNulls[i] = resultNulls[i]->asMutable<uint64_t>();
+    }
+
+    // Final Row Vector
+    const auto finalArraySize = rows.end();
+    BufferPtr finalArrayIndices = allocateIndices(finalArraySize, pool);
+    BufferPtr finalArrayNulls =
+        AlignedBuffer::allocate<bool>(finalArraySize, pool, bits::kNotNull);
+    BufferPtr finalArrayOffsets = allocateOffsets(finalArraySize, pool);
+    BufferPtr finalArraySizes = allocateSizes(finalArraySize, pool);
+
+    auto rawFinalArrayOffsets = finalArrayOffsets->asMutable<vector_size_t>();
+    auto rawFinalArraySizes = finalArraySizes->asMutable<vector_size_t>();
+
+    // Create right offsets/indexes for the individual and final result arrays.
+    int rowIndex = 0;
+    int finalResultIndex = 0;
+    rows.applyToSelected([&](auto row) {
+      // Get the max size for that row.
+      auto maxRowSize = maxArraySizes[row];
+      rawFinalArrayOffsets[finalResultIndex] = rowIndex;
+
+      for (int arr = 0; arr < numInputArrays; arr++) {
+        auto arrayOffset = rawOffsets[arr][indices[arr][row]];
+        auto vectorSize = rawSizes[arr][indices[arr][row]];
+        for (int i = 0; i < maxRowSize; i++) {
```

Comment:
This loop is to fill in first 'size' elements with sequential values started with 'offset' and the rest with nulls, right? Consider doing these two operations in bulk.

```
std::iota(rawResultIndices[i] + rowIndex, rawResultIndices[i] + rowIndex + size, offset);
bits::fillBits(rawResultNulls[i], rowIndex + size, rowIndex + maxArraySize, kNull);
```

Reply:
Very Nice. Thank you!